Math/src/main/java/org/tribuo/math/neighbour/kdtree/DimensionNode.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.tribuo.math.neighbour.kdtree;
+
+import org.tribuo.math.distance.DistanceType;
+import org.tribuo.math.la.SGDVector;
+
+import static org.tribuo.math.neighbour.kdtree.KDTree.IntAndVector;
+import static org.tribuo.math.neighbour.kdtree.KDTree.DistanceIntAndVectorBoundedMinHeap;
+
+/**
+ * A node used in a k-d tree {@link KDTree}. A node is a point from a dataset and is placed according to its value
+ * at a specific dimension of the point.
+ */
+public final class DimensionNode {
```

Comment:
Thinking about this more, does this class (and its methods) need to be public? Might be nicer to have it as an implementation detail, that way we have the flexibility to change it later.

Reply:
Yes, locking this down is a better approach here.